modules/admin_manual/pages/configuration/server/security/hsmdaemon/index.adoc

```diff
@@ -0,0 +1,501 @@
+The communication happens via {unix-sockets-url}[UNIX sockets] or {network-sockets-url}[TCP sockets] and is authorized by a shared token that the daemon stores in the ownCloud database via a REST/JSON route.
+
+ownCloud internally uses OpenSSL to en-/decrypt keys and needs to be extended to support en-/decrypt operations via the new daemon. 
+The current solution encrypts the ownCloud master key with a key from the HSM. 
+
+NOTE: From the technical point of view the `Crypt` class is extended to handle the key generation in the HSM device and also to get the key from HSM. 
+For the read/write operation on a file, the request goes to the HSM and then, based on the keys fetched from HSM, the files are encrypted or decrypted. 
+The keys are not replaced.
+
+== How The HSM Daemon Interacts with ownCloud
+
+Upon startup, the daemon will generate a token and send it to ownCloud via a new REST/JSON route. 
+After connecting with the HSM daemon, an unsophisticated, line-based, protocol is used (every line ends with CRLF):
+
+. ownCloud sends the token read from database.
+. The daemon compares the received token with its token and returns an `"OK"` line.
+. ownCloud then sends the data it wants to decrypt as a {base64-encoding-url}[Base64-encoded], one-line string.
+. The daemon returns the decrypted data as a Base64-encoded one-line string.
+
+Doing so ensures that an evil admin will need to wiretap the communication between either the database or the HSM daemon and ownCloud.
+
+== Quick Overview
+
+HSM support consists of two core parts:
+
+. An actual HSM PKCS11 module.
+. An hsmdaemon that provides a {jwt-url}[JWT]-protected web API for the PKCS11 stack to generate key pairs and decrypt data.
+
+== Deployment Recommendation
+
+We recommend running hsmdaemon on every web server to reduce latency. 
+
+== Installation
+
+Integrating the hsmdaemon with ownCloud requires 3 steps; these are:
+
+. xref:install-a-pkcs11-module[Install a PKCS11 Module]
+. xref:install-and-configure-the-hsmdaemon[Install and Configure the hsmdaemon]
+. xref:configure-owncloud[Configure ownCloud]
+
+[NOTE]
+====
+The installation instructions in this guide have been designed to work with xref:installation/system_requirements.adoc#server[ownCloud's supported operating systems].
+If you are using a different operating system or distribution, please adjust the instructions to suit your environment.
+====
+
+=== Install a PKCS11 Module
+
+==== Install Using a Preconfigured PKCS11 Module
+
+At least one PKCS11 library is necessary, which is typically provided by an HSM vendor. 
+However, if none is available, you can xref:configuration/server/security/hsmdaemon/install-softhsm2.adoc[install a software HSM (SoftHSM2)].
+SoftHSM2 is the most simple approach to employ storage encryption with a master key in an HSM. 
+This is because it hands off the master key material to a user other than the web server user.
```

Comment:
That's not the reason for being the "most simple approach". It's a benefit compared to not using an HSM at all. 
=> The point here is that using softhsm is not at all as secure as using an actual **hardware** device. But still there's this benefit that another user on the system is used for the hsmdaemon.

Reply:
SoftHSM2 would only give the upper edge when compared with oC instance without HSM.
How about rephrasing this sentence as:
```
SoftHSM2 is an approach which user could opt, when no HSM hardware is available. It would simulate the same procedure how a hardware device would behave. Like creating key pairs, providing the key to decrypt the file of oC instance when oC user tries to view the file etc.
```

The sentence structure might not be accurate ( pardon me for that ). Let me know if this looks ok?

Reply:
I've been playing around with that paragraph and have come up with the following, based on the existing information. Can you review it and let me know if it's correct?

> SoftHSM2 can also substitute for an HSM when one is not available. For example, when creating key pairs, SoftHSM2 can provide the key to decrypt the file of an ownCloud instance, when the ownCloud user tries to view the file.

I'm not confident on the accuracy of the last sentence.

Reply:
Looks fine to me.

Reply:
> For example, when creating key pairs, SoftHSM2 can provide the key to decrypt the file of an ownCloud instance, when the ownCloud user tries to view the file.

I don't understand the above. I think the point is:
- Storage encryption with master key on the storage => master key is located on oC storage, next to the actual data, accessible by apache web server user
- Storage encryption with master key in softhsm => master key is located on within softhsm on any storage or machine of your choice, accessible only by softhsm user on that machine
- Storage encryption with master key in an actual HSM => master key is located in a dedicated hardware device, no way to access it

Please verify this with @butonic ^

Reply:
ping @butonic

Reply:
Since this is security related documentation half truths help no one. I learned that any security related documentation that does not define threat models is suspicious. So what are we protecting against?
* normal master (or user) key based encryption: files are stored on an external storage service. keys reside in the owncloud server, the storage admin only sees garbage. This scenario actually makes sense if storage and ownCloud services are managed by different people.

* using hsmdaemon (with softhsm2) to store the keys instead of ownCloud itself. Keys are stored with a different owner, preventing apache from directly reading the key material. This might help when someone gains access to the apache user and can read files. It won't help if he can manipulate php files and run arbitrary code. Still ... an improvement. But I find it hard to justify.

* using hsmdaemon with an actual HSM: keys are stored outside the machine, data cannot be accessed when someone steals the complete machine and boots it outside of the corporate network: the hsm is not accessible: files cannot be read. This makes sense.

* You can actually run hsmdaemon with softhsm2 on a different machine in the network, but IIRC traffic is not yet encrypted, so you need to setup your own secure sh tunnel before doing that. This would be the poor mans network attached HSM ... actually this might be a cool blog post. Or we invest in encrypting hsmdaemon traffic. However, I would prefer moving it to reva instead.

Now, regarding the above sentences: there is no difference between softhsm2 and any other HSM. ownCloud encrypts all files with a symmetric file individual key. To read a file that key has to be decrypted using our hsmdaemon. The hsmdeamon uses pkcs11 to talk to whatever hsm driver (module) is configured. Inside the HSM the master key has been generated and It should not leave the HSM, although there might be process reasons to allow that. The real question is what is the threat model and related to that: does the driver / module store the key on the same machine?

Back to the original phrasing:
```
At least one PKCS11 library is necessary, which is typically provided by an HSM vendor. 
However, if none is available, you can xref:configuration/server/security/hsmdaemon/install-softhsm2.adoc[install a software HSM (SoftHSM2)].
SoftHSM2 is the most simple approach to employ storage encryption with a master key in an HSM. 
This is because it hands off the master key material to a user other than the web server user.
```

The first sentence is there to stress that it is not our responsibility to come up with a pkcs11 driver for an HSM. That is the responsibility of the system admin.
The second sentence is intended to point to an example. SoftHSM2 can be used because documentation is freely available. Other HSM vendors tend to make their docs hard to find.
The third and fourth sentence give a reason for using SoftHSM2 over the core encryption, because it is relatively easy to set up and already increases security.

All of it falls under `==== Install Using a Preconfigured PKCS11 Module`

Maybe we should just make a note out of the last three sentences, leaving `At least one PKCS11 library is necessary, which is typically provided by an HSM vendor.` as the main point. Softhsm2 actually only serves an as example...